### PR TITLE
Scorm events

### PR DESCRIPTION
**Pull request by jlowe64**
Expanding events to handle scorm submissions.

### CODE REVIEW

**Comment by jlowe64** [comment]
In progress, sorry for the mess. :)

**Comment by jlowe64** [comment]
@garemoko I always get a bit confused about what should be the object and what should be the context. Also, this one gets slightly more confusing as in the scorm_scoes_track db, there could be 100 rows associated with one activity.

Any hints would be appreciated...

**Comment by garemoko** [comment]
The object includes anything relating to the thing being acted on, in this case the SCO. The object is always exactly the same in every single statement about that object. 

The context contains stuff about the overall action being taken and might vary between statements. 

E.g. where a SCO is located, the device used to access it etc. all form part of the context. The name and description of the SCO belong in the object. 

Regarding scorm_scoes_track, one big difference from SCORM to xAPI is that SCORM deals with statuses, whilst xAPI deals with events. The point of translation therefore needs to be changes in status. So if the SCO reports completion is true, when it was previously false then that's a statement. If it reports completion is true but it was already true, then no statement is required. 

The following may be helpful:

https://tincanapi.com/scorm-evolved
https://github.com/adlnet/xAPI-SCORM-Profile
https://github.com/adlnet/SCORM-to-xAPI-Wrapper

File: src/Events/ScormScorerawSubmitted.php [hunk 1]
```diff
@@ -14,7 +14,7 @@ class ScormScorerawSubmitted extends Event {
     public function read(array $opts) {
         return array_merge_recursive(parent::read($opts), [
             'verb' => [
-                'id' => 'http://adlnet.gov/expapi/verbs/completed',
+                'id' => 'http://adlnet.gov/expapi/verbs/' . $opts['scorm_status'], // Make this verb match whatever comes through
```
Comment (ryansmith94) [inline src/Events/ScormScorerawSubmitted.php]: We can't really just do that. We need to create a new method to select a verb based on the scorm status.

Comment (jlowe64) [inline src/Events/ScormScorerawSubmitted.php]: Can do.

**Comment by jlowe64** [comment]
@ryansmith94 @garemoko Ready for review when you have a moment. I'm ignoring the issue that's cropped up for codeclimate because I need to evaluate the overall impact of changing that and make sure it's done right.

File: src/Events/Event.php [hunk 2]
```diff
@@ -148,4 +148,49 @@ protected function readVerbDisplay($opts) {
         $lang = isset(static::$verb_display[$lang]) ? $lang : array_keys(static::$verb_display)[0];
         return [$lang => static::$verb_display[$lang]];
     }
+
```
Comment (ryansmith94) [inline src/Events/Event.php]: We should probably extract these methods (readScormVerb and readScormScoes) into a ScormEvent class that can be extended instead.

File: src/Events/ScormScorerawSubmitted.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
+<?php namespace XREmitter\Events;
+
+class ScormScorerawSubmitted extends Event {
```
Comment (ryansmith94) [inline src/Events/ScormScorerawSubmitted.php]: Should probably be renamed to ScormScoreRawSubmitted (with a capital "R" in "Raw").

File: tests/EventTest.php [hunk 4]
```diff
@@ -128,6 +128,25 @@ protected function constructDiscussion() {
         ];
     }
 
+    protected function constructScormTracking() {
```
Comment (ryansmith94) [inline tests/EventTest.php]: Same thing here as above, could extract methods into a ScormEventTest.

File: src/Controller.php [hunk 5]
```diff
@@ -21,7 +21,9 @@ class Controller extends PhpObj {
         'scorm_launched' => 'ScormLaunched',
         'training_session_attend' => 'Attended',
         'training_session_enrol' => 'EventEnrol',
-        'training_session_unenrol' => 'EventUnenrol'
+        'training_session_unenrol' => 'EventUnenrol',
+        'scorm_scoreraw_submitted' => 'ScormScorerawSubmitted',
```
Comment (garemoko) [inline src/Controller.php]: You don't have this recipe in the translator

File: src/Events/Event.php [hunk 6]
```diff
@@ -148,4 +148,49 @@ protected function readVerbDisplay($opts) {
...
+    protected function readScormVerb($opts) {
+        $scormStatus = $opts['scorm_status'];
+        $verbBaseUrl = 'http://adlnet.gov/expapi/verbs/';
+        $verb = array();
+
+        switch ($scormStatus) {
+            case 'failed':
+                $verbUrl = $verbBaseUrl . $scormStatus;
```
Comment (garemoko) [inline src/Events/Event.php]: creating IRIs directly from database values seems like a bad idea

File: docs/examples/ScormScorerawSubmitted.json [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    "timestamp": "2015-01-01T01:00Z",
+    "verb": {
+        "id": "http:\/\/adlnet.gov\/expapi\/verbs\/completed",
+        "display": {
+            "en": "completed"
+        }
+    },
+    "result": {
+        "score": {
+            "raw": 100,
```
Comment (garemoko) [inline docs/examples/ScormScorerawSubmitted.json]: I don't think this should be hard coded!!!

File: src/Events/ScormScorerawSubmitted.php [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+    /**
+     * Reads data for an event.
+     * @param [String => Mixed] $opts
+     * @return [String => Mixed]
+     * @override Event
+     */
+    public function read(array $opts) {
+        return array_merge_recursive(parent::read($opts), [
+            'verb' => $this->readScormVerb($opts),
```
Comment (garemoko) [inline src/Events/ScormScorerawSubmitted.php]: Will this always be completed?

File: src/Events/Event.php (hunk 6 above)
Comment (jlowe64) [inline src/Events/Event.php]: @garemoko what alternative would you suggest?

File: src/Events/ScormScorerawSubmitted.php (hunk 8 above)
Comment (jlowe64) [inline src/Events/ScormScorerawSubmitted.php]: @garemoko I'll take a look at this.

File: tests/EventTest.php (hunk 4 above)
Comment (jlowe64) [inline tests/EventTest.php]: @ryansmith94 will do

File: src/Events/Event.php (hunk 6 above)
Comment (garemoko) [inline src/Events/Event.php]: Use a fixed list of verbs and a switch, if/else structure or array key to choose them.

**Comment by davidpesce** [comment]
@garemoko and the last one

**Comment by garemoko** [comment]
@davidpesce can you ask me some specific questions? I'm not clear on what you're asking me to do on these.

**Comment by davidpesce** [comment]
This is passing the tests, but I'd still like to verify the verb is being passed correctly.

**Pull request merged**